src/utils/user_registry.ts

```diff
@@ -0,0 +1,89 @@
+import {
+  IResponseErrorForbiddenNotAuthorized,
+  IResponseErrorInternal,
+  IResponseErrorValidation,
+  ResponseErrorForbiddenNotAuthorized,
+  ResponseErrorValidation
+} from "@pagopa/ts-commons/lib/responses";
+import { FiscalCode } from "@pagopa/ts-commons/lib/strings";
+import { toError } from "fp-ts/lib/Either";
+import { none, Option, some } from "fp-ts/lib/Option";
+import { taskEither } from "fp-ts/lib/TaskEither";
+import { fromEither, fromLeft, tryCatch } from "fp-ts/lib/TaskEither";
+import { User } from "../../generated/userregistry-api/User";
+
+import { UserRegistryAPIClient } from "../clients/userregistry_client";
+import { errorsToError, toResponseErrorInternal } from "./conversions";
+
+export const getUserId = (
+  apiClient: ReturnType<UserRegistryAPIClient>,
+  externalId: FiscalCode
+) =>
+  tryCatch(() => apiClient.getUserIdByExternalId({ externalId }), toError)
+    .mapLeft<IResponseErrorInternal | IResponseErrorForbiddenNotAuthorized>(
+      toResponseErrorInternal
+    )
+    // Validation (Either) -> taskEither
+    .chain(_ =>
+      fromEither(_).mapLeft(errs =>
+        toResponseErrorInternal(errorsToError(errs))
+      )
+    )
+    .chain<Option<{ id: string }>>(res => {
+      switch (res.status) {
+        case 200:
+          return taskEither.of(some(res.value));
+        case 404:
+          return taskEither.of(none);
+        default:
+          return fromLeft(ResponseErrorForbiddenNotAuthorized);
+      }
+    });
+
+export const postUser = (
+  apiClient: ReturnType<UserRegistryAPIClient>,
+  user: User
+) => {
+  return tryCatch(
+    () =>
+      apiClient.createUser({
+        body: {
+          ...user
+        }
+      }),
+    toError
+  )
+    .mapLeft<IResponseErrorInternal | IResponseErrorValidation>(
+      toResponseErrorInternal
+    )
+    .chain(_ =>
+      fromEither(_).mapLeft(errs =>
+        toResponseErrorInternal(errorsToError(errs))
+      )
+    )
+    .chain<User>(res =>
+      res.status === 201
+        ? taskEither.of(res.value)
+        : fromLeft(ResponseErrorValidation("Bad Input", res.value.detail))
+    );
+};
+
+export const blurUser = (
+  apiClient: ReturnType<UserRegistryAPIClient>,
+  user: User,
+  fiscalCode: FiscalCode
+) => {
+  return getUserId(apiClient, fiscalCode)
+    .mapLeft(error => {
+      return toResponseErrorInternal(toError(error));
+    })
+    .chain(maybeUserID => {
+      return maybeUserID.isSome()
+        ? taskEither.of(maybeUserID.toUndefined())
+        : postUser(apiClient, user)
+            .mapLeft(err => toResponseErrorInternal(toError(err)))
+            .map(u => {
+              return { id: u.id };
+            });
+    });
+};
```

Comment:
I should prefer return an Option in this case and let the caller decide how it can be used. With this implementation the return type of this chain should be `TaskEither<IResponseErrorInternal, { id: string } | undefined>`

Reply:
Fix with 3175de2